# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/gc/Marking.cpp
# Commit: 2fa53cd9ac4e
# Full Hash: 2fa53cd9ac4e821763dbc4402e58b58184ce230b
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2022-03-04 04:18:28
# Regressor Bug: 1643586
# File Overlap Count: 1
# Description:
#   Bug 1643586 - Part 6: Stop using JSContext for GC helper thread tasks r=jandem
#   
#   This patch moves the debug-only GC state from JSContext to JSFreeOp. Support
#   functions for this move to FreeOp.h and most RAII classes move to GCInternals.h.
#   
# ==============================================================================

diff -r 054a08d13999 -r 2fa53cd9ac4e js/src/gc/Marking.cpp
--- a/js/src/gc/Marking.cpp	Thu Mar 03 17:49:18 2022 +0000
+++ b/js/src/gc/Marking.cpp	Thu Mar 03 17:49:18 2022 +0000
@@ -208,7 +208,7 @@
   bool isUnmarkGrayTracer = IsTracerKind(trc, JS::TracerKind::UnmarkGray);
   bool isClearEdgesTracer = IsTracerKind(trc, JS::TracerKind::ClearEdges);
 
-  if (TlsContext.get()->isMainThreadContext()) {
+  if (TlsContext.get()) {
     // If we're on the main thread we must have access to the runtime and zone.
     MOZ_ASSERT(CurrentThreadCanAccessRuntime(rt));
     MOZ_ASSERT(CurrentThreadCanAccessZone(zone));
@@ -2745,13 +2745,12 @@
 
   // Allow the current thread access if it is sweeping or in sweep-marking, but
   // try to check the zone. Some threads have access to all zones when sweeping.
-  JSContext* cx = TlsContext.get();
-  MOZ_ASSERT(cx->gcUse != JSContext::GCUse::Finalizing);
-  if (cx->gcUse == JSContext::GCUse::Sweeping ||
-      cx->gcUse == JSContext::GCUse::Marking) {
+  JSFreeOp* fop = TlsFreeOp.get();
+  MOZ_ASSERT(fop->gcUse() != GCUse::Finalizing);
+  if (fop->gcUse() == GCUse::Sweeping || fop->gcUse() == GCUse::Marking) {
     Zone* zone = thing->zoneFromAnyThread();
-    MOZ_ASSERT_IF(cx->gcSweepZone,
-                  cx->gcSweepZone == zone || zone->isAtomsZone());
+    MOZ_ASSERT_IF(fop->gcSweepZone(),
+                  fop->gcSweepZone() == zone || zone->isAtomsZone());
     return;
   }
 
@@ -2798,7 +2797,7 @@
   // Permanent things are never finalized by non-owning runtimes. Zone state is
   // unknown in this case.
 #ifdef DEBUG
-  JSRuntime* rt = TlsContext.get()->runtime();
+  JSRuntime* rt = TlsFreeOp.get()->runtimeFromAnyThread();
   MOZ_ASSERT_IF(IsOwnedByOtherRuntime(rt, thing), thing->isMarkedBlack());
 #endif
 
@@ -3167,9 +3166,11 @@
              CurrentThreadIsGCMarking());
   MOZ_ASSERT(markColor() == MarkColor::Black);
 
-  AutoGeckoProfilerEntry profilingStackFrame(
-      TlsContext.get(), "GCMarker::traceBarrieredCells",
-      JS::ProfilingCategoryPair::GCCC_Barrier);
+  mozilla::Maybe<AutoGeckoProfilerEntry> profilingStackFrame;
+  if (JSContext* cx = TlsContext.get()) {
+    profilingStackFrame.emplace(cx, "GCMarker::traceBarrieredCells",
+                                JS::ProfilingCategoryPair::GCCC_Barrier);
+  }
 
   BarrierBuffer& buffer = barrierBuffer();
   while (!buffer.empty()) {